webtau-graphql/src/main/java/org/testingisdocumenting/webtau/graphql/model/GraphQLRequest.java

```diff
@@ -18,15 +18,90 @@
 
 import org.testingisdocumenting.webtau.http.json.JsonRequestBody;
 import org.testingisdocumenting.webtau.http.request.HttpRequestBody;
+import org.testingisdocumenting.webtau.utils.JsonParseException;
+import org.testingisdocumenting.webtau.utils.JsonUtils;
 import org.testingisdocumenting.webtau.utils.StringUtils;
 
 import java.util.HashMap;
 import java.util.Map;
+import java.util.Optional;
 
 import static org.testingisdocumenting.webtau.utils.CollectionUtils.notNullOrEmpty;
 
 public class GraphQLRequest {
-    public static HttpRequestBody body(String query, Map<String, Object> variables, String operationName) {
+    private final String query;
+    private final Map<String, Object> variables;
+    private final String operationName;
+
+    public GraphQLRequest(String query) {
+        this(query, null, null);
+    }
+
+    public GraphQLRequest(String query, Map<String, Object> variables, String operationName) {
+        this.query = query;
+        this.variables = variables;
+        this.operationName = operationName;
+    }
+
+    public static Optional<GraphQLRequest> fromHttpRequest(String method, String url, HttpRequestBody requestBody) {
+        if (!"POST".equals(method) || !"/graphql".equals(url) || !(requestBody instanceof JsonRequestBody)) {
+            return Optional.empty();
+        }
+
+        Map<String, ?> request;
+        try {
+            request = JsonUtils.deserializeAsMap(requestBody.asString());
+        } catch (JsonParseException ignore) {
+            // Ignoring as it's not a graphql request
+            return Optional.empty();
+        }
+
+        if (!request.containsKey("query")) {
+            // Ignoring as it's not a graphql request
+            return Optional.empty();
+        }
+
+        Object queryObj = request.get("query");
+        if (!(queryObj instanceof String)) {
+            // Ignoring as it's not a graphql request
+            return Optional.empty();
+        }
+        String query = (String) queryObj;
+
+        Map<String, Object> variables = null;
+        Object variablesObj = request.get("variables");
+        if (variablesObj instanceof Map) {
+            variables = (Map<String, Object>) variablesObj;
+        } else if (variablesObj != null) {
+            // Ignoring as it's not a graphql request
+            return Optional.empty();
+        }
+
+        String operationName = null;
```

Comment:
should we move these block as creation functions?

Reply:
There was a reason why I didn't do it but I can't remember what it is.  I'll have another look though.

Reply:
OK, I remember now.  There are two reasons I didn't stick it into functions:
1. behaviour for query is different to vars/operationName as the query is mandatory and the others are not.
2. there are three cases for vars/operationName.  It's null, it's not the expected type or it is the expected type.  Null and expected type result in me proceeding, not expected type results in not creating a GraphQLRequest as it's not a valid GraphQL request.  I think in all likelihood the usage of a function to deal with this will look almost identical to the current code and as the actual checks for type and null are very simple it didn't seem worthwhile deduplicating those parts alone.